### PR TITLE
[fix] Copy only app missing in new apps dir

### PR DESCRIPTION
**Pull request by zamentur**
## Problem
https://github.com/YunoHost-Apps/nextcloud_ynh/issues/168

## Solution
I suggest to use the old method use in owncloud package
https://github.com/YunoHost-Apps/owncloud_ynh/blob/master/scripts/upgrade#L92

## PR Status
- [x] Code finished.
- [ ] Tested with Package_check.
- [x] Fix or enhancement tested.
- [x] Upgrade from last version tested.
- [x] Can be reviewed and tested.

## Validation
---
*Minor decision*
- **Upgrade previous version** : 
- [x] **Code review** : Maniack C
- [x] **Approval (LGTM)** : JimboJoe
- [x] **Approval (LGTM)** : Kay0u
- **CI succeeded** : 
[Build Status](https://ci-apps-hq.yunohost.org/jenkins/job/nextcloud_ynh%20PR170/)  
When the PR is marked as ready to merge, you have to wait for 3 days before really merging it.

### CODE REVIEW

File: scripts/upgrade [hunk 1]
```diff
@@ -205,7 +205,10 @@ do
     # Backup 3rd party applications from the current nextcloud
     # But do not overwrite if there is any upgrade
     # (apps directory already exists in Nextcloud archive)
-    cp -a --update "$final_path/apps" "$tmpdir"
+    for nc_app_dir in $(sudo ls "${final_path}/apps"); do
```
Comment (maniackcrudelis) [inline scripts/upgrade]: Do you need sudo ?

File: scripts/upgrade [hunk 2]
```diff
@@ -205,7 +205,10 @@ do
...
+       [[ ! -d "${tmpdir}/apps/${nc_app_dir}" ]] \
```
Comment (maniackcrudelis) [inline scripts/upgrade]: Use simple bracket when you don't need double ones please.

**Comment by zamentur** [comment]
Note: a user on yunohost support chat had the same issue.

**Comment by alexAubin** [comment]
>Note: a user on yunohost support chat had the same issue.

And it took 3 hours to debug / fix it :|

File: scripts/upgrade (hunk 1 above)
Comment (zamentur) [reply scripts/upgrade]: no i removed it

**Comment by JimboJoe** [comment]
This PR still needs some love...

**Comment by maniackcrudelis** [comment]
Is a simple cp not enough !?
Never had that error, I don't really know what's the real problem, nor why we're doing that for Nextcloud.

**Comment by JimboJoe** [comment]
Now reconsidering this, has anyone ever tried the other way around: delete existing "core" apps before copying the new version...? :thinking: 
Wouldn't that be safer if the app isn't compatible any more with the new Nextcloud version (and provided it's included in the Nextcloud archive)...?

**Comment by maniackcrudelis** [comment]
Or simply overwrite everything, as we do usually. So we would be sure to have the last version.

**Comment by JimboJoe** [comment]
Well, I think the original problem for this PR was that there was a mix between files of older and newer version of the apps, so the idea is to have consistent apps.

**Comment by JimboJoe** [comment]
CI failed:
```
Warning: yunohost.hook <lambda> - [6288.1] cp: failed to access '/tmp/tmp.s50gkDHPBG/apps/activity'$'n''admin_audit'$'n''comments'$'n''dav'$'n''encryption'$'n''federatedfilesharing'$'n''federation'$'n''files'$'n''files_external'$'n''files_pdfviewer'$'n''files_sharing'$'n''files_texteditor'$'n''files_trashbin'$'n''files_versions'$'n''files_videoplayer'$'n''firstrunwizard'$'n''gallery'$'n''logreader'$'n''lookup_server_connector'$'n''nextcloud_announcements'$'n''notifications'$'n''oauth2'$'n''password_policy'$'n''provisioning_api'$'n''serverinfo'$'n''sharebymail'$'n''survey_client'$'n''systemtags'$'n''theming'$'n''twofactor_backupcodes'$'n''updatenotification'$'n''user_external'$'n''user_ldap'$'n''workflowengine': File name too long`
```

**Comment by maniackcrudelis** [comment]
Shouldn't it be a ` ls -1` instead of a simple `ls` ?

**Comment by JimboJoe** [comment]
Indeed, works much better!

After restoring the previous apps, the `occ upgrade` command looks for updates for all apps in the appstore, so I think this change handles the "mixed versions" problems and the rest will be handled smoothly.

In favour of merging this quickly into testing.

**Comment by JimboJoe** [comment]
Sorry for the duplicate and thanks for the fix. Still wondering how I happened to do that... :sweat_smile:

**Comment by maniackcrudelis** [comment]
Code is clean, agree on the idea. But when merged, should be tested in real situations.

**Comment by JimboJoe** [comment]
My production server has some apps which are not up-to-date, so I'll try the testing and see how it goes under the hood.

**Comment by maniackcrudelis** [comment]
Still not working...
```
Warning: yunohost.hook <lambda> - [6291.1] cp: failed to access '/tmp/tmp.stJCPcj8ol/apps/activity'$'n''admin_audit'$'n''comments'$'n''dav'$'n''encryption'$'n''federatedfilesharing'$'n''federation'$'n''files'$'n''files_external'$'n''files_pdfviewer'$'n''files_sharing'$'n''files_texteditor'$'n''files_trashbin'$'n''files_versions'$'n''files_videoplayer'$'n''firstrunwizard'$'n''gallery'$'n''logreader'$'n''lookup_server_connector'$'n''nextcloud_announcements'$'n''notifications'$'n''oauth2'$'n''password_policy'$'n''provisioning_api'$'n''serverinfo'$'n''sharebymail'$'n''survey_client'$'n''systemtags'$'n''theming'$'n''twofactor_backupcodes'$'n''updatenotification'$'n''user_external'$'n''user_ldap'$'n''workflowengine': File name too long
```

**Comment by maniackcrudelis** [comment]
Seems to work as expected.
```
2019-03-23 21:30:55,447 DEBUG    yunohost.hook <lambda> - [6242.1] + cd /var/www/nextcloud/apps
2019-03-23 21:30:55,447 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,448 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/activity/ ']'
2019-03-23 21:30:55,448 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,448 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/admin_audit/ ']'
2019-03-23 21:30:55,449 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,449 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/comments/ ']'
2019-03-23 21:30:55,449 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,450 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/dav/ ']'
2019-03-23 21:30:55,450 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,450 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/encryption/ ']'
2019-03-23 21:30:55,451 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,451 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/federatedfilesharing/ ']'
2019-03-23 21:30:55,452 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,452 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/federation/ ']'
2019-03-23 21:30:55,452 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,453 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/files/ ']'
2019-03-23 21:30:55,453 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,453 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/files_external/ ']'
2019-03-23 21:30:55,454 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,455 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/files_pdfviewer/ ']'
2019-03-23 21:30:55,455 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,455 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/files_sharing/ ']'
2019-03-23 21:30:55,456 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,456 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/files_texteditor/ ']'
2019-03-23 21:30:55,456 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,457 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/files_trashbin/ ']'
2019-03-23 21:30:55,457 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,457 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/files_versions/ ']'
2019-03-23 21:30:55,458 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,458 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/files_videoplayer/ ']'
2019-03-23 21:30:55,459 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,459 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/firstrunwizard/ ']'
2019-03-23 21:30:55,459 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,460 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/gallery/ ']'
2019-03-23 21:30:55,460 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,460 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/logreader/ ']'
2019-03-23 21:30:55,461 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,461 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/lookup_server_connector/ ']'
2019-03-23 21:30:55,461 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,462 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/nextcloud_announcements/ ']'
2019-03-23 21:30:55,462 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,462 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/notifications/ ']'
2019-03-23 21:30:55,463 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,463 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/oauth2/ ']'
2019-03-23 21:30:55,463 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,464 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/password_policy/ ']'
2019-03-23 21:30:55,464 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,464 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/provisioning_api/ ']'
2019-03-23 21:30:55,465 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,465 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/serverinfo/ ']'
2019-03-23 21:30:55,465 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,466 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/sharebymail/ ']'
2019-03-23 21:30:55,466 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,467 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/survey_client/ ']'
2019-03-23 21:30:55,467 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,467 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/systemtags/ ']'
2019-03-23 21:30:55,468 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,468 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/theming/ ']'
2019-03-23 21:30:55,468 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,469 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/twofactor_backupcodes/ ']'
2019-03-23 21:30:55,469 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,469 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/updatenotification/ ']'
2019-03-23 21:30:55,470 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,470 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/user_external/ ']'
2019-03-23 21:30:55,471 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,471 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/user_ldap/ ']'
2019-03-23 21:30:55,471 DEBUG    yunohost.hook <lambda> - [6242.1] + for nc_app_dir in */
2019-03-23 21:30:55,472 DEBUG    yunohost.hook <lambda> - [6242.1] + '[' '!' -d /tmp/tmp.PXbv4IV6R2/apps/workflowengine/ ']'
```

**Comment by JimboJoe** [comment]
Looks like I *really* tested it this time :stuck_out_tongue_closed_eyes:

**Comment by maniackcrudelis** [comment]
Don't say that, people will know we're not testing our code ;)

**Comment by JimboJoe** [comment]
Tester c'est douter

**Comment by JimboJoe** [comment]
Need for a new review from @YunoHost-Apps/apps-group (and for the CI!)

**Comment by kay0u** [comment]
Can be merged in 3 days

**Comment by JimboJoe** [comment]
Ouch, we should have released the testing before merging that one... So, I'll advertise a new testing, and that will need a few more days/weeks of testing.

**Comment by alexAubin** [comment]
Uuuuh but this is from Feb 26 and has already been tested by several maintainers

I also saw a user like 1 week ago encountering the very issue fixed by this PR ...  I literally made him test this PR (which was painful but he already knew git clones and git checkout.

I don't know how many more people we can expect to test this specific issue which only happens during upgrade :/ ...

**Comment by JimboJoe** [comment]
The new testing release revealed a bug (not linked to that correction, though)...

**Pull request merged**